src/csharp/Grpc.Core.Tests/Internal/AsyncCallTest.cs

```diff
@@ -493,6 +493,20 @@ public void ServerStreaming_StartCallFailureDoesntLeakResources()
             Assert.IsTrue(fakeCall.IsDisposed);
         }
 
+        [Test]
+        public void ServerStreaming_ServerResponseFailureDoesntLeakResources()
+        {
+            // Test for https://github.com/grpc/grpc/issues/28153
+            // Memory leak if resources not released when server unavailable
+            asyncCall.StartServerStreamingCall("request1");
+
+            // error reading response headers - possibly because server unavailable
+            fakeCall.ReceivedResponseHeadersCallback.OnReceivedResponseHeaders(false, new Metadata());
```

Comment:
since OnReceivedResponseHeaders can be triggered anytime after the call has started, the user can be attempting to read a message while that happens (e.g just before or just after) - it would be good to test those scenarios and make sure the expected behavior doesn't change.

Basically the ordering of OnReceivedResponseHeader(), OnReceivedStatusOnClient() and MoveNext() happening is fully at random (the two events can be delivered in any order and MoveNext() is invoked by the user so we have no control over when exactly it happens).